File: Minimize the Heights I - GFG/minimize-the-heights-i.cpp
// { Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
//User function Template for C++

class Solution{   
public:
    int getMinDiff(int arr[], int n, int k) {
        vector<pair<int,int>> vec;
        
        for(int i=0;i<n;i++){
            vec.push_back(make_pair(arr[i]-k,i));
            vec.push_back(make_pair(arr[i]+k,i));
        }
        
        sort(vec.begin(),vec.end());
        int visit[n];
        fill(visit,visit+n,0);
        int s=0,e=0,c=0,ans=vec[vec.size()-1].first-vec[0].first;
        while(e<2*n){
            
            while(e<2*n && c<n){
                pair<int,int> temp=vec[e];
                int num=temp.first;
                int idx=temp.second;
                if(visit[idx]==0){
                    c++;
                }
                visit[idx]++;
                e++;
            }
            if(c<n) break;
            while(c==n){
                pair<int,int> temp=vec[s];
                int num=temp.first;
                int idx=temp.second;
                visit[idx]--;
                if(visit[idx]==0){
                    c--;
                }
                ans=min(ans,vec[e-1].first-vec[s].first);
                s++;
            }
            
            // cout<<s<<" "<<e<<endl;
        }
        return ans;
    }
};

// { Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, k;
        cin >> k;
        cin >> n;
        int arr[n];
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        Solution ob;
        auto ans = ob.getMinDiff(arr, n, k);
        cout << ans << "\n";
    }
    return 0;
}  // } Driver Code Ends

File: Minimize the Heights I - GFG/minimize-the-heights-i.java
// { Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine().trim());
        while (tc-- > 0) {
            String[] inputLine;
            inputLine = br.readLine().trim().split(" ");
            int k = Integer.parseInt(inputLine[0]);
            inputLine = br.readLine().trim().split(" ");
            int n = Integer.parseInt(inputLine[0]);
            int[] arr = new int[n];
            inputLine = br.readLine().trim().split(" ");
            for (int i = 0; i < n; i++) {
                arr[i] = Integer.parseInt(inputLine[i]);
            }

            int ans = new Solution().getMinDiff(arr, n, k);
            System.out.println(ans);
        }
    }
}// } Driver Code Ends


//User function Template for Java

class Solution {
    int getMinDiff(int[] arr, int n, int k) {
        
        int hash[]=new int[n];
        int c=0;
        List<int []> list=new ArrayList<>();
        
        for(int i=0;i<n;i++){
            
            list.add(new int[]{i,arr[i]+k});
            list.add(new int[]{i,arr[i]-k});
        }
        Collections.sort(list, new Comparator<>(){
            public int compare(int a[],int b[]){
                return a[1]-b[1];
            }
        });
        
        int i=0,s=i,ans=Integer.MAX_VALUE,nn=list.size();
        while(i<nn){
            
            while(c<n && i<nn){
                int aa[]=list.get(i);
                hash[aa[0]]++;
                if(hash[aa[0]]==1) c++;
                i++;
            }
            if(c!=n) break;
            while(s<i){
                int aa[]=list.get(s++);
                ans=Math.min(ans,list.get(i-1)[1]-aa[1]);
                hash[aa[0]]--;
                if(hash[aa[0]]==0){
                    c--;
                    break;
                }
            }
        }
        return ans;
    }
}
